File: src/com/megamind/algorithm/SortingHelper.java
package com.megamind.algorithm;

import com.megamind.algorithm.insertion_sort.InsertionSortKt;
import com.megamind.algorithm.selection_sort.SelectionSortKt;

import java.math.BigDecimal;
import java.math.RoundingMode;

import static com.megamind.algorithm.merge_sort.MergeSortKt.mergeSort;

public class SortingHelper {
    public static <E extends Comparable<E>> boolean isSorted(E[] arr) {
        for (int i = 1; i < arr.length; i++) {
            if (arr[i - 1].compareTo(arr[i]) > 0) {
                return false;
            }
        }
        return true;
    }

    public static <E> void swap(E[] array, int i, int j) {
        E t = array[i];
        array[i] = array[j];
        array[j] = t;
    }

    public static <E extends Comparable<E>> void sortTest(String sortName, E[] arr) {
        long startTime = System.nanoTime();
        switch (sortName) {
            case "selectionSort": {
                SelectionSortKt.selectionSort(arr);
                break;
            }
            case "selectionSort2": {
                SelectionSortKt.selectionSort2(arr);
                break;
            }
            case "insertionSort": {
                InsertionSortKt.insertionSort(arr);
                break;
            }
            case "insertionSort2": {
                InsertionSortKt.insertionSort2(arr);
                break;
            }
            case "insertionSort3": {
                InsertionSortKt.insertionSort3(arr);
                break;
            }
            case "mergeSort": {
                mergeSort(arr);
                break;
            }
            default:
                break;
        }
        long endTime = System.nanoTime();
        BigDecimal duration = new BigDecimal((endTime - startTime) / 1000_000_000.0).setScale(6, RoundingMode.HALF_UP);
        if (!isSorted(arr)) {
            throw new RuntimeException(sortName + " failed");
        }
        System.out.println(sortName + " n = " + arr.length + ", time: " + duration + " s");
    }
}

File: src/com/megamind/algorithm/SortingHelper.kt
package com.megamind.algorithm

import com.megamind.algorithm.insertion_sort.insertionSort
import com.megamind.algorithm.insertion_sort.insertionSort2
import com.megamind.algorithm.insertion_sort.insertionSort3
import com.megamind.algorithm.selection_sort.selectionSort
import com.megamind.algorithm.selection_sort.selectionSort2
import java.lang.RuntimeException
import java.math.BigDecimal
import java.math.RoundingMode

fun <T : Comparable<T>> isSorted(arr: Array<T>): Boolean {
    for (i in 1 until arr.size) {
        if (arr[i - 1] > arr[i])
            return false
    }
    return true
}

fun <T : Comparable<T>> sortTest(sortName: String, arr: Array<T>) {
    val startTime = System.nanoTime()
    when (sortName) {
        "selectionSort" -> {
            selectionSort(arr)
        }
        "selectionSort2" -> {
            selectionSort2(arr)
        }
        "insertionSort" -> {
            insertionSort(arr)
        }
        "insertionSort2" -> {
            insertionSort2(arr)
        }
        "insertionSort3" -> {
            insertionSort3(arr)
        }
    }
    val endTime = System.nanoTime()
    val duration = BigDecimal((endTime - startTime) / 1000_000_000.0).setScale(6, RoundingMode.HALF_UP)
    if (!isSorted(arr)) {
        throw RuntimeException("$sortName failed")
    }
    println("$sortName, n = ${arr.size}, time: $duration s")
}

fun <T> swap(arr: Array<T>, i: Int, j: Int) {
    val t = arr[i]
    arr[i] = arr[j]
    arr[j] = t
}